plugin/evm/vm_validators_state_test.go

```diff
@@ -0,0 +1,334 @@
+package evm
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/database/memdb"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow"
+	commonEng "github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/snow/engine/enginetest"
+	avagoValidators "github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/snow/validators/validatorstest"
+	"github.com/ava-labs/subnet-evm/core"
+	"github.com/ava-labs/subnet-evm/plugin/evm/validators"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+)
+
+func TestValidatorState(t *testing.T) {
+	require := require.New(t)
+	genesis := &core.Genesis{}
+	require.NoError(genesis.UnmarshalJSON([]byte(genesisJSONLatest)))
+	genesisJSON, err := genesis.MarshalJSON()
+	require.NoError(err)
+
+	vm := &VM{}
+	ctx, dbManager, genesisBytes, issuer, _ := setupGenesis(t, string(genesisJSON))
+	appSender := &enginetest.Sender{T: t}
+	appSender.CantSendAppGossip = true
+	testNodeIDs := []ids.NodeID{
+		ids.GenerateTestNodeID(),
+		ids.GenerateTestNodeID(),
+		ids.GenerateTestNodeID(),
+	}
+	testValidationIDs := []ids.ID{
+		ids.GenerateTestID(),
+		ids.GenerateTestID(),
+		ids.GenerateTestID(),
+	}
+	ctx.ValidatorState = &validatorstest.State{
+		GetCurrentValidatorSetF: func(ctx context.Context, subnetID ids.ID) (map[ids.ID]*avagoValidators.GetCurrentValidatorOutput, uint64, bool, error) {
+			return map[ids.ID]*avagoValidators.GetCurrentValidatorOutput{
+				testValidationIDs[0]: {
+					NodeID:    testNodeIDs[0],
+					PublicKey: nil,
+					Weight:    1,
+				},
+				testValidationIDs[1]: {
+					NodeID:    testNodeIDs[1],
+					PublicKey: nil,
+					Weight:    1,
+				},
+				testValidationIDs[2]: {
+					NodeID:    testNodeIDs[2],
+					PublicKey: nil,
+					Weight:    1,
+				},
+			}, 0, false, nil
+		},
+	}
+	appSender.SendAppGossipF = func(context.Context, commonEng.SendConfig, []byte) error { return nil }
+	err = vm.Initialize(
+		context.Background(),
+		ctx,
+		dbManager,
+		genesisBytes,
+		[]byte(""),
+		[]byte(""),
+		issuer,
+		[]*commonEng.Fx{},
+		appSender,
+	)
+	require.NoError(err, "error initializing GenesisVM")
+
+	// Test case 1: state should not be populated until bootstrapped
+	require.Equal(0, vm.validatorState.GetValidationIDs().Len())
+	_, _, err = vm.uptimeManager.CalculateUptime(testNodeIDs[0])
+	require.ErrorIs(database.ErrNotFound, err)
+	require.False(vm.uptimeManager.StartedTracking())
+
+	// Test case 2: state should be populated after bootstrapped
+	require.NoError(vm.SetState(context.Background(), snow.Bootstrapping))
```

Comment:
should we set the state to Bootstrapping before checking that it's not populated, such as to line 80? (i.e., only gets populated in NormalOps)

Reply:
I think VM is in bootstrapping state by default, but yeah let's make it more explicit.